From 17d47e9868cb89e2d4a34125b53687b8efbd5d2b Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Thu, 12 Dec 2013 16:58:01 +0700
Subject: Dalvik: BinaryExpression builder does not accept MIR with 3 uses

BZ: 158559

BinaryExpression builder does not expect MIR with 3 uses while
it corresponds to ushr-long instruction.
The patch adds this case.

Change-Id: Ibcfd0883102e4597560010bdd923d137ead3bc08
Category: device enablement
Domain: AOSP.Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, need rework
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/Expression.cpp |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/Expression.cpp b/vm/compiler/Expression.cpp
index a3e61d0..b958685 100644
--- a/vm/compiler/Expression.cpp
+++ b/vm/compiler/Expression.cpp
@@ -339,7 +339,8 @@ BinaryExpression * BinaryExpression::mirToExpression (MIR * mir,
     //scenarios we are dealing with:
     // 1. One use means one operand is non-wide VR and other is literal
     // 2. Two uses means both operands are non-wide VRs
-    // 3. Four uses means both operands are wide VRs
+    // 3. Three uses means first operand is wide VRs while the second one is non-wide VR
+    // 4. Four uses means both operands are wide VRs
 
     if (ssaRep->numUses == 1)
     {
@@ -357,6 +358,14 @@ BinaryExpression * BinaryExpression::mirToExpression (MIR * mir,
         lhs = VirtualRegister::convertToVR (vrToExpression, ssaRep->uses[0]);
         rhs = VirtualRegister::convertToVR (vrToExpression, ssaRep->uses[1]);
     }
+    else if (ssaRep->numUses == 3)
+    {
+        assert (ssaRep->uses != 0);
+
+        lhs = VirtualRegister::convertToVR (vrToExpression, ssaRep->uses[0],
+                ssaRep->uses[1], true);
+        rhs = VirtualRegister::convertToVR (vrToExpression, ssaRep->uses[2]);
+    }
     else if (ssaRep->numUses == 4)
     {
         assert (ssaRep->uses != 0);
-- 
1.7.4.1

